src/main/java/org/traccar/WebDataHandler.java

```diff
@@ -155,32 +181,95 @@ public String formatRequest(Position position) throws UnsupportedEncodingExcepti
+                    throw new RuntimeException("Forwarding formatting error", e);
+                }
+
+                requestBuilder = client.target(formattedUrl).request();
+                if (header != null && !header.isEmpty()) {
+                    for (String line: header.split("\\r?\\n")) {
+                        String[] values = line.split(":", 2);
+                        requestBuilder.header(values[0].trim(), values[1].trim());
+                    }
+                }
+
+                if (json) {
+                    payload = prepareJsonPayload(position);
+                }
+
+                deliveryPendingCurrent.incrementAndGet();
+
+                send();
             }
-        }
 
-        Invocation.Builder requestBuilder = client.target(url).request();
+            private void send() {
+                if (json) {
+                    requestBuilder.async().post(Entity.json(payload), this);
+                } else {
+                    requestBuilder.async().get(this);
+                }
+            }
 
-        if (header != null && !header.isEmpty()) {
-            for (String line: header.split("\\r?\\n")) {
-                String[] values = line.split(":", 2);
-                requestBuilder.header(values[0].trim(), values[1].trim());
+            private void retry() {
+                try {
+                    String message = "Position forwarding failed.";
+                    if (!retryEnabled) {
+                        LOGGER.warn(message);
+                    } else {
+                        int pending = deliveryPendingCurrent.get();
+                        if (pending <= deliveryPendingLimit) {
+                            LOGGER.warn(message + " Pending: " + pending
+                                + ". Retrying in " + delay + " seconds.");
+                            GlobalTimer.getTimer().newTimeout(new TimerTask() {
+                                @Override
+                                public void run(Timeout timeout) {
+                                    if (!timeout.isCancelled()) {
+                                        if (delay < retryDelayMax) {
+                                            delay++;
+                                        }
+                                        send();
+                                    }
+                                }
+                            }, delay, TimeUnit.SECONDS);
+                            return;
+                        }
+                        LOGGER.warn(message + " Pending: " + pending
+                            + ". Delivery will not be retried.");
+                    }
+                } catch (Exception e) {
+                }
+                deliveryPendingCurrent.decrementAndGet();
+            }
+
+            @Override
+            public void completed(Response response) {
+                if (response.getStatus() == 200) {
```

Comment:
All 2xx codes should be a success. There might be a helper method somewhere to check it.

Reply:
You were absolutely right. Done!